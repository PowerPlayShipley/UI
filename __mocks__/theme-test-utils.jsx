// TODO: Fix this??

import * as React from 'react'
import userEvent from '@testing-library/user-event'
import { render as rtlRender } from '@testing-library/react'

import { ThemeProvider } from "styled-components";

import theme from '../src/themes'
import GlobalStyle from "../src/components/GlobalStyle";

function renderWithTheme(ui, { ...options } = {}) {
  const Wrapper = ({ children }) => (
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
      <div id='modal-root' />
    </ThemeProvider>
  )

  return rtlRender(ui, { wrapper: Wrapper, ...options })
}

// THE BELLOW COME FROM `https://github.com/styled-components/styled-components/blob/0aa3170c255a49cd41c3fbeb2b8051b5d132f229/src/test/utils.js`

const stripComments = (str) => str.replace(/\/\*.*?\*\/\n?/g, '');

const stripWhitespace = (str) => str
  .trim()
  .replace(/([;\{\}])/g, '$1  ')
  .replace(/\s+/g, ' ');

const getCSS = (scope) =>
  Array.from(scope.querySelectorAll('style'))
    .map(tag => tag.innerHTML)
    .join('\n')
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

function expectCSSMatches ( _expectation, opts = { ignoreWhitespace: true }) {
  // NOTE: This should normalise both CSS strings to make irrelevant mismatches less likely
  const expectation = _expectation
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

  const css = getCSS(document);

  if (opts.ignoreWhitespace) {
    const stripped = stripWhitespace(stripComments(css));
    expect(stripped).toEqual(stripWhitespace(expectation));
    return stripped;
  } else {
    expect(css).toEqual(expectation);
    return css;
  }
};

export * from '@testing-library/react';

export { renderWithTheme, expectCSSMatches };
