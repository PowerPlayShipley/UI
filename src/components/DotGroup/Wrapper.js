import styled from "styled-components";

const Wrapper = styled.span`
  display: flex;
  justify-content: flex-start;
  align-items: center;
  
  span {
    margin-right: 4px;
  }

  span:last-child {
    margin-right: 0;
  }
`

export default Wrapper
