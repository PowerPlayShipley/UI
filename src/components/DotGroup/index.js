import React, { useMemo } from 'react'
import clsx from 'clsx'
import PropTypes from 'prop-types'

import Wrapper from './Wrapper'

import Dot from '../Dot'

import tooltip from "../../utils/tooltip";

function DotGroup ({ className, tooltip: tt, dots, size = 'md', ...rest }) {
  // Map to the correct object
  const dotObject = useMemo(() => {
    return dots.map((el) => typeof el === 'string' ? ({ color: el, size }) : el)
  }, [dots, size])

  return (
    <Wrapper className={clsx(className)} {...tooltip(tt)} {...rest}>{
      dotObject.map((el, idx) => (
        <Dot key={idx} data-testid='dotitem' {...el} />
      ))}
    </Wrapper>
  )
}

DotGroup.propTypes = {
  // Add a grouped tooltip
  tooltip: PropTypes.object,
  dots: PropTypes.arrayOf(
    PropTypes.oneOfType([
      PropTypes.exact({
        size: PropTypes.oneOf([
          'sm', 'md', 'lg'
        ]),
        color: PropTypes.string
      }),
      PropTypes.string
    ])
  ).isRequired,
  // Allow for a generalised size when passing colors as the dots proptype
  size: PropTypes.oneOf([
    'sm', 'md', 'lg'
  ])
}

export default DotGroup
