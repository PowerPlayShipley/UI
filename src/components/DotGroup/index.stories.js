import DotGroup from "./index";

export default {
  title: 'Dot/Group',
  component: DotGroup
}

const Template = (args) => (<DotGroup {...args} />);

export const Default = Template.bind({});
Default.args = {
  dots: ['red', 'green', 'blue'],
  size: 'md',
  tooltip: {
    tooltip: 'Includes Tooltip'
  }
}
