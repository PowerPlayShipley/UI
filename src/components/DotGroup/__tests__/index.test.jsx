/**
 * @jest-environment jsdom
 */

import React from 'react'
import { renderWithTheme } from '../../../../__mocks__/theme-test-utils'

import DotGroup from "../index";

describe('Display Dot', () => {
  it('should render dot', async function () {
    const { findAllByTestId } = renderWithTheme(<DotGroup dots={['red', 'green', 'blue']} />)
    expect(await findAllByTestId('dotitem')).toHaveLength(3)
  });

  it('should create valid snapshot', () => {
    const { container } = renderWithTheme(<DotGroup dots={['red', 'green', 'blue']} />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
