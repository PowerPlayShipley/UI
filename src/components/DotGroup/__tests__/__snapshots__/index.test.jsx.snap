// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display Dot should create valid snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 span {
  margin-right: 4px;
}

.c0 span:last-child {
  margin-right: 0;
}

.c1 {
  height: 20px;
  width: 20px;
  background-color: red;
  border-radius: 50%;
  display: inline-block;
}

.c2 {
  height: 20px;
  width: 20px;
  background-color: green;
  border-radius: 50%;
  display: inline-block;
}

.c3 {
  height: 20px;
  width: 20px;
  background-color: blue;
  border-radius: 50%;
  display: inline-block;
}

<span
  class="c0"
>
  <span
    class="c1"
    color="red"
    data-testid="dotitem"
  />
  <span
    class="c2"
    color="green"
    data-testid="dotitem"
  />
  <span
    class="c3"
    color="blue"
    data-testid="dotitem"
  />
</span>
`;
